include/solvers/multiphysics_interface.h

```diff
@@ -0,0 +1,461 @@
+   *
+   */
+  MultiphysicsInterface(
+    const SimulationParameters<dim> &nsparam,
+    std::shared_ptr<parallel::DistributedTriangulationBase<dim>>
+                                       p_triangulation,
+    std::shared_ptr<SimulationControl> p_simulation_control);
+
+  std::vector<PhysicsID>
+  get_active_physics()
+  {
+    return active_physics;
+  }
+
+  /**
+   * @brief Call for the solution of all physics
+   *
+   * @param time_stepping_method Time-Stepping method with which the assembly is called
+   */
+  void
+  solve(const Parameters::SimulationControl::TimeSteppingMethod
+                   time_stepping_method,
+        const bool force_matrix_renewal)
+  {
+    for (auto &iphys : physics)
+      {
+        solve_physics(iphys.first, time_stepping_method, force_matrix_renewal);
+      }
+    for (auto &iphys : block_physics)
+      {
+        solve_block_physics(iphys.first,
+                            time_stepping_method,
+                            force_matrix_renewal);
+      }
+  }
+
+
+  /**
+   * @brief Call for the solution of a single physics
+   *
+   * @param time_stepping_method Time-Stepping method with which the assembly is called
+   */
+  void
+  solve_physics(const PhysicsID physics_id,
+                const Parameters::SimulationControl::TimeSteppingMethod
+                           time_stepping_method,
+                const bool force_matrix_renewal)
+  {
+    AssertThrow(std::find(active_physics.begin(),
+                          active_physics.end(),
+                          physics_id) != active_physics.end(),
+                ExcInternalError());
+
+    physics[physics_id]->solve_non_linear_system(time_stepping_method,
+                                                 false,
+                                                 force_matrix_renewal);
+  }
+
+  /**
+   * @brief Call for the solution of a single block physics
+   *
+   * @param time_stepping_method Time-Stepping method with which the assembly is called
+   */
+  void
+  solve_block_physics(const PhysicsID physics_id,
+                      const Parameters::SimulationControl::TimeSteppingMethod
+                                 time_stepping_method,
+                      const bool force_matrix_renewal)
+  {
+    AssertThrow(std::find(active_physics.begin(),
+                          active_physics.end(),
+                          physics_id) != active_physics.end(),
+                ExcInternalError());
+
+    block_physics[physics_id]->solve_non_linear_system(time_stepping_method,
+                                                       false,
+                                                       force_matrix_renewal);
+  }
+
+
+  /**
+   * @brief Call the attachment of the solution vector to the data out for enabled
+   * auxiliary physics.
+   */
+  void
+  attach_solution_to_output(DataOut<dim> &data_out)
+
+  {
+    for (auto &iphys : physics)
+      {
+        iphys.second->attach_solution_to_output(data_out);
```

Comment:
Iphys is a std::pair. So a map works by creating a list of pairs which are key/value. It's a random access structure (instead of having a vector). Pairs are accessed by .first and .second so when you loop over the elements of a map in a linear fashion, you get pairs one by one.

Reply:
I get the std::map and std::pair part (I read about it in the course), but I am confused why is Iphys a pair and not a map. On l. 443 `physics` is defined with :
`std::map<PhysicsID, std::shared_ptr<AuxiliaryPhysics<dim, TrilinosWrappers::MPI::Vector>>> physics`

So `physics` is a map, with `physics->first` being the PhysicsID and `physics->second` the shared pointers of auxiliary physics. As `iphys` is a reference to `physics`, why is it a pair, rather than a map as well?

Anyway I feel it could be good to document it, otherwise the reader would have to go to l.443 to get what this ->first and ->second are exactly (you could put other type of key in your map so it is not really implicit here I think)

Reply:
When you loop over the elements of a map sequentially, what you get is a list of all the key/element pair. So iphys in this case is an std::pair. It takes the value of each key/pair combination one by one. It is how you loop sequentially through a map (which is not the regular way maps are accessed generally).
I'll add some documentation to it.